manual/src/cmds/runtime-tracing.etex

```diff
@@ -314,4 +322,146 @@ Percentile 	 Latency (ms)
 99.9990 	 9.17
 99.9999 	 9.17
 100.0000 	 9.17
-\end{verbatim}
\ No newline at end of file
+\end{verbatim}
+
+\section{s:runtime-tracing-custom-events}{Custom events}
+
+\subsection{s:runtime-tracing-custom-events-overview}{Overview}
+
+The runtime events system supports defining custom events. Identified by a
+globally unique name, they are emitted with payloads of built-in (\texttt{unit},
+\texttt{int}, \texttt{span}) and user-defined types. To understand the
+manipulation of custom events, it is useful to know how they are transported and
+stored: their representation consists of a name string (in fact, an index into
+an array of all custom names) and an arbitrary array of bytes. Custom event
+types can be \emph{registered} by providing encoding and decoding functions to
+and from an array of bytes, via the function
+\texttt{Runtime_events..Type.register}.
+
+Defining a new custom event (whether its payload is of a built-in type or a
+custom type) is done via another registration function,
+\texttt{Runtime_events.User.register}. This function records the association
+between the custom event's name, its type, and a \emph{tag}. The tag is then
+used when emitting or consuming custom events; it acts as a shorthand for the
+event name.
+
+Once an event is registered, values for that event are emitted using
+\texttt{Runtime_events.User.write}.
+
+To summarize, for a user to emit and consume custom events using a custom type
+they need to:
+
+\begin{enumerate}
+\item register the custom type
+\item extend the \texttt{Runtime_events.User.tag} variant with a new constructor
+\item register the new custom event, binding together the custom event name, the
+  new tag, and the new custom type
+\item emit instances of the event using \texttt{Runtime_events.User.write}
+\item indicate a callback which should receive all events of the custom
+  type using \texttt{Runtime_events.Callbacks.add_user_event} to register it
+  inside a \texttt{Runtime_events.Callbacks.t} object to be used when polling
+  (see section~\ref{s:runtime-tracing-ocaml-apis} above).
+\item \emph{(Optional, if there are distinct events registered with the same
+  custom type:)} In that callback, pattern-match on the event tag in order to
+  discriminate between different event tags of the same custom type.
+\end{enumerate}
+
+Note that if the emitter and the consumer are different programs, both must
+perform steps 1 to 3 to register custom events and custom event types (if any).
```

Comment:
If the emitter and consumer are different programs, can they actually declare their tags independently by using the same `type Runtime_events.User.tag += Foo` line? I would worry that two tags end up having different representations despite having the same name, and that this would not work -- unless they link a shared module that declares the tag.

My understanding reading the implementation is that in fact the tag are never exchanged between event producers and event consumers, only the names are, so tags are just there for convenience and there is no need to use the same tag on both sides.

Reply:
That is correct. The real identifier is the name.

Reply:
I think that this should be clarified in the manual, because that sounds like a question (do the producer and consumer need to use the *same* tags?) that is going to confuse most people familiar with extensible variants.

Reply:
Good idea, I added a sentence to that effect in fd83647e5a.